File: iEPG2iCal/Classes/iEPG/TVProgramInfo.m
//
//  TVProgramInfo.m
//  iEPG2iCal
//
//  Created by SUZUKI, Wataru on 5/1/11.
//  Copyright 2011 sz50.com. All rights reserved.
//

#import "TVProgramInfo.h"


@implementation TVProgramInfo

static char CRLF[2] = {0xd, 0xa};

@synthesize dateTimeEnd;
@synthesize dateTimeStart;
@synthesize genre;
@synthesize memo;
@synthesize performer;
@synthesize programSubtitle;
@synthesize programTitle;
@synthesize station;
@synthesize subGenre;


#pragma -

+ (id)tvProgramInfoWithData:(NSData *)data {
    return [[[self alloc] initWithData:data] autorelease];
}

#pragma -

- (id)init {
    self = [super init];
    if (self) {
        calendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSGregorianCalendar];
        encoding = NSASCIIStringEncoding;
    }
    
    return self;
}

- (id)initWithData:(NSData *)d {
    self = [self init];
    if (self) {
        data = [d retain];
        [self parse];
    }
    
    return self;
}


#pragma -

- (void)parse {
    NSData *CRLFData = [NSData dataWithBytes:CRLF length:2];
    BOOL headerPartEnds = NO;
    NSUInteger flags = NSYearCalendarUnit | NSMonthCalendarUnit | NSDayCalendarUnit;
    NSDateComponents *dateComponentsStart = [calendar components:flags fromDate:[NSDate date]];
    NSDateComponents *dateComponentsEnd   = [calendar components:flags fromDate:[NSDate date]];
    NSUInteger cursor = 0;
    for (; cursor < data.length;) {
        NSRange range1 = [data rangeOfData:CRLFData options:0 range:NSMakeRange(cursor, data.length - cursor)];
        
        if (range1.location == NSNotFound) {
            break;
        }
        
        if (cursor == range1.location) {
            headerPartEnds = YES;
        }
        
        if (!headerPartEnds) {
            NSString *line = [[[NSString alloc] initWithData:[data subdataWithRange:NSMakeRange(cursor, range1.location - cursor)] encoding:encoding] autorelease];
            NSRange range2 = [line rangeOfString:@":"];
            
            NSString *name = [[line substringToIndex:range2.location] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
            NSString *value = [[line substringFromIndex:range2.location + range2.length] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
            
            if ([name isCaseInsensitiveLike:@"content-type"]) {
                [self parseContentType:value];
                
            } else if ([name isCaseInsensitiveLike:@"version"]) {
                version = [value retain];
                
            } else if ([name isCaseInsensitiveLike:@"station"]) {
                self.station = value;
                
            } else if ([name isCaseInsensitiveLike:@"year"]) {
                [dateComponentsStart setYear:[value intValue]];
                [dateComponentsEnd setYear:[value intValue]];
                
            } else if ([name isCaseInsensitiveLike:@"month"]) {
                [dateComponentsStart setMonth:[value intValue]];
                [dateComponentsEnd setMonth:[value intValue]];
                
            } else if ([name isCaseInsensitiveLike:@"date"]) {
                [dateComponentsStart setDay:[value intValue]];
                [dateComponentsEnd setDay:[value intValue]];
                
            } else if ([name isCaseInsensitiveLike:@"start"]) {
                NSArray *array = [value componentsSeparatedByString:@":"];
                NSUInteger hour = [[array objectAtIndex:0] intValue];
                NSUInteger minute = [[array objectAtIndex:1] intValue];
                if (hour > 24) {
                    [dateComponentsStart setDay:dateComponentsStart.day + 1];
                    hour -= 24;
                }
                [dateComponentsStart setHour:hour];
                [dateComponentsStart setMinute:minute];
                
            } else if ([name isCaseInsensitiveLike:@"end"]) {
                NSArray *array = [value componentsSeparatedByString:@":"];
                NSUInteger hour = [[array objectAtIndex:0] intValue];
                NSUInteger minute = [[array objectAtIndex:1] intValue];
                if (hour > 24) {
                    [dateComponentsEnd setDay:dateComponentsEnd.day + 1];
                    hour -= 24;
                }
                [dateComponentsEnd setHour:hour];
                [dateComponentsEnd setMinute:minute];
                
            } else if ([name isCaseInsensitiveLike:@"program-title"]) {
                self.programTitle = value;
                
            } else if ([name isCaseInsensitiveLike:@"program-subtitle"]) {
                self.programSubtitle = value;
                
            } else if ([name isCaseInsensitiveLike:@"performer"]) {
                self.performer = value;
                
            } else if ([name isCaseInsensitiveLike:@"genre"]) {
                self.genre = [NSNumber numberWithInt:[value intValue]];
                
            } else if ([name isCaseInsensitiveLike:@"subgenre"]) {
                self.subGenre = [NSNumber numberWithInt:[value intValue]];
                
            }
        } else {
            cursor += CRLFData.length;
            self.memo = [[[NSString alloc] initWithData:[data subdataWithRange:NSMakeRange(cursor, data.length - cursor)] encoding:encoding] autorelease];
            break;
        }
        cursor = range1.location + CRLFData.length;
    }
    
    self.dateTimeStart = [calendar dateFromComponents:dateComponentsStart];
    self.dateTimeEnd = [calendar dateFromComponents:dateComponentsEnd];
    if ([self.dateTimeStart compare:self.dateTimeEnd] == NSOrderedDescending) {
        self.dateTimeEnd = [self.dateTimeEnd dateByAddingTimeInterval:60 * 60 * 24];
    }
}

- (void)parseContentType:(NSString *)value {
    NSArray *components = [value componentsSeparatedByString:@";"];
    
    for (NSUInteger i = 0; i < components.count; i++) {
        if (i > 0) {
            NSString *s     = [components objectAtIndex:i];
            NSRange  range  = [s rangeOfString:@"="];
            NSString *name  = [[s substringToIndex:range.location] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
            NSString *value = [[s substringFromIndex:range.location + range.length] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
            
            if ([name isCaseInsensitiveLike:@"charset"]) {
                if ([value isCaseInsensitiveLike:@"UTF-8"]) {
                    encoding = NSUTF8StringEncoding;
                } else if ([value isCaseInsensitiveLike:@"Shift_JIS"]) {
                    encoding = NSShiftJISStringEncoding;
                }
            }
        }
    }
}


#pragma - 

- (NSString *)description {
    NSMutableString *s = [NSMutableString string];
    [s appendString:@"{\n"];
    [s appendFormat:@"        version: %@\n", version];
    [s appendFormat:@"        station: %@\n", self.station];
    [s appendFormat:@"start date time: %@\n", self.dateTimeStart.description];
    [s appendFormat:@"  end date time: %@\n", self.dateTimeEnd.description];
    [s appendFormat:@"          title: %@\n", self.programTitle];
    [s appendFormat:@"       subtitle: %@\n", self.programSubtitle];
    [s appendFormat:@"      performer: %@\n", self.performer];
    [s appendFormat:@"          genre: %@ - %@\n", self.genre, self.subGenre];
    [s appendFormat:@"           memo: %@\n", self.memo];
    [s appendString:@"}\n"];
    return s;
}


#pragma - 

- (void)dealloc {
    [calendar release];
    [data release];
    [dateTimeEnd release];
    [dateTimeStart release];
    [genre release];
    [memo release];
    [performer release];
    [programSubtitle release];
    [programTitle release];
    [station release];
    [subGenre release];
    [version release];
    
    calendar        = nil;
    data            = nil;
    dateTimeEnd     = nil;
    dateTimeStart   = nil;
    genre           = nil;
    memo            = nil;
    performer       = nil;
    programSubtitle = nil;
    programTitle    = nil;
    station         = nil;
    subGenre        = nil;
    version         = nil;
    
    [super dealloc];
}

@end

File: iEPG2iCal 2/Classes/iEPG/TVProgramInfo.swift
//
//  TVProgramInfo.swift
//  iEPG2iCal 2
//
//  Created by Wataru SUZUKI on 8/3/14.
//  Copyright (c) 2014 sz50.com. All rights reserved.
//

import Cocoa

class TVProgramInfo {
    let CRLF: NSData = NSData(bytes: "\r\n".UTF8String, length: 2)

    var calendar:      NSCalendar
    var data:          NSData?
    var encoding:      NSStringEncoding
    var endDateTime:   NSDate?
    var genre:         NSNumber?
    var memo:          NSString?
    var performer:     NSString?
    var startDateTime: NSDate?
    var station:       NSString?
    var subGenre:      NSNumber?
    var subtitle:      NSString?
    var title:         NSString?
    var version:       NSString?

    init(let data: NSData) {
        self.calendar = NSCalendar(calendarIdentifier: NSGregorianCalendar)
        self.encoding = NSASCIIStringEncoding
        self.data     = data

        self.parse()
    }

    func parse() {
        let calendarUnitFlags: NSCalendarUnit = .YearCalendarUnit | .MonthCalendarUnit | .DayCalendarUnit

        let dateComponentsStart: NSDateComponents = self.calendar.components(calendarUnitFlags, fromDate: NSDate())
        let dateComponentsEnd:   NSDateComponents = self.calendar.components(calendarUnitFlags, fromDate: NSDate())

        var headerPartEnds: Bool = false
        var cursor:         Int  = 0
        while cursor < self.data?.length {
            let range1: NSRange = self.data!.rangeOfData(CRLF, options:NSDataSearchOptions(0), range:NSMakeRange(cursor, self.data!.length - cursor))

            if range1.location == NSNotFound {
                break
            }

            headerPartEnds = cursor == range1.location

            if (!headerPartEnds) {
                let lineData: NSData   = self.data!.subdataWithRange(NSMakeRange(cursor, range1.location - cursor))
                let line:     NSString = NSString(data: lineData, encoding: encoding)
                let range2:   NSRange  = line.rangeOfString(":")

                let name:  NSString = line.substringToIndex(range2.location).stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceCharacterSet())
                let value: NSString = line.substringFromIndex(range2.location + range2.length).stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceCharacterSet())

                switch name {
                case let name where name.isCaseInsensitiveLike("content-type"):
                    (_, self.encoding) = Utils.parseContentType(value)

                case let name where name .isCaseInsensitiveLike("version"):

                    self.version = value
                case let name where name.isCaseInsensitiveLike("station"):
                    self.station = value

                case let name where name.isCaseInsensitiveLike("year"):
                    dateComponentsStart.setValue(value.integerValue, forComponent: NSCalendarUnit.YearCalendarUnit)
                    dateComponentsEnd.setValue(value.integerValue, forComponent: NSCalendarUnit.YearCalendarUnit)

                case let name where name.isCaseInsensitiveLike("month"):
                    dateComponentsStart.setValue(value.integerValue, forComponent: NSCalendarUnit.MonthCalendarUnit)
                    dateComponentsEnd.setValue(value.integerValue, forComponent: NSCalendarUnit.MonthCalendarUnit)

                case let name where name.isCaseInsensitiveLike("date"):
                    dateComponentsStart.setValue(value.integerValue, forComponent: NSCalendarUnit.DayCalendarUnit)
                    dateComponentsEnd.setValue(value.integerValue, forComponent: NSCalendarUnit.DayCalendarUnit)

                case let name where name.isCaseInsensitiveLike("start"):
                    self.setTime(value, dateComponents: dateComponentsStart)

                case let name where name.isCaseInsensitiveLike("end"):
                    self.setTime(value, dateComponents: dateComponentsEnd)

                case let name where name.isCaseInsensitiveLike("program-title"):
                    self.title = value

                case let name where name.isCaseInsensitiveLike("program-subtitle"):
                    self.subtitle = value

                case let name where name.isCaseInsensitiveLike("performer"):
                    self.performer = value

                case let name where name.isCaseInsensitiveLike("genre"):
                    self.genre = NSNumber(integer: value.integerValue)
                    
                case let name where name.isCaseInsensitiveLike("subgenre"):
                    self.subGenre = NSNumber(integer: value.integerValue)
                    
                default:
                    break
                }
                
            } else {
                cursor += CRLF.length
                self.memo = NSString(data: self.data!.subdataWithRange(NSMakeRange(cursor, self.data!.length - cursor)), encoding: self.encoding)
                break
            }
            cursor = range1.location + CRLF.length
        }

        self.startDateTime = self.calendar.dateFromComponents(dateComponentsStart)
        self.endDateTime   = self.calendar.dateFromComponents(dateComponentsEnd)

        if (self.startDateTime!.compare(self.endDateTime!) == NSComparisonResult.OrderedDescending) {
            self.endDateTime = self.endDateTime?.dateByAddingTimeInterval(NSTimeInterval(60 * 60 * 24))
        }
    }

    func setTime(let string: String, let dateComponents: NSDateComponents) {
        let array: NSArray = string.componentsSeparatedByString(":")

        var hour:   Int = array.objectAtIndex(0).integerValue
        var minute: Int = array.objectAtIndex(1).integerValue

        if (hour > 24) {
            dateComponents.setValue(dateComponents.day + 1, forComponent: NSCalendarUnit.DayCalendarUnit)
            hour -= 24
        }

        dateComponents.setValue(hour, forComponent: NSCalendarUnit.HourCalendarUnit)
        dateComponents.setValue(minute, forComponent: NSCalendarUnit.MinuteCalendarUnit)
    }

}
